azure_functions_worker/extension.py

```diff
@@ -0,0 +1,276 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+from types import ModuleType
+from typing import Any, Callable, List, Optional
+import sys
+import importlib
+import logging
+import functools
+from .utils.common import is_python_version
+from .utils.wrappers import enable_feature_by
+from .constants import (
+    SYSTEM_LOG_PREFIX,
+    PYTHON_ISOLATE_WORKER_DEPENDENCIES,
+    PYTHON_ENABLE_WORKER_EXTENSIONS,
+    PYTHON_ENABLE_WORKER_EXTENSIONS_DEFAULT,
+    PYTHON_ENABLE_WORKER_EXTENSIONS_DEFAULT_39,
+    FUNC_EXT_POST_FUNCTION_LOAD,
+    FUNC_EXT_PRE_INVOCATION,
+    FUNC_EXT_POST_INVOCATION,
+    APP_EXT_POST_FUNCTION_LOAD,
+    APP_EXT_PRE_INVOCATION,
+    APP_EXT_POST_INVOCATION
+)
+from .logging import logger
+
+
+class ExtensionManager:
+    """This marks if the ExtensionManager has already proceeded a detection,
+    if so, the sdk will be cached in .extension_enabled_sdk
+    """
+    is_sdk_detected: bool = False
+
+    """This is a cache of azure.functions module that supports extension
+    interfaces. If this is None, that mean the sdk does not support extension.
+    """
+    extension_enabled_sdk: Optional[ModuleType] = None
```

Comment:
Can `is_sdk_detected` and `extension_enabled_sdk` be private?

Reply:
Also the methods such as `get_sdk_from_sys_path`, `is_extension_enabled_in_sdk`, `get_sdk_version` etc.? If yes, then please make it private as I don't see any other class needing access to these methods.

Reply:
Yeah this is a nice catch, otherwise the extension manager is exposing too many interfaces.
I will only expose the essential ones `function_load_extension`, `invocation_extension`, `get_invocation_wrapper`, and `get_invocation_wrapper_async` and make all others private.